### ISSUE TITLE
Improve positioning of Ulauncher

### ISSUE THREAD

**Issue by FinnHering** [Ulauncher/Ulauncher#1043]
## Issue
**disclaimer:** This issue is really niece, and I don't expect anyone other than myself to work on improving this.

Right now, Ulauncher appears in the center of the currently focused _hardware_ display. This is probably a great solution 99.9999% of the time. 

But as soon as you split up your _hardware_ monitor(s) into multiple 'X-monitors' this behavior becomes annoying. 

Let me visualize what I mean (I'm starting Ulauncher while the right 'X-monitor' is focused).

Current behavior:
![grafik](https://user-images.githubusercontent.com/92150718/170740414-4e61fec3-cad4-436a-bb9b-5dd9081b880c.png)

Wanted behavior:
![grafik](https://user-images.githubusercontent.com/92150718/170740691-5cafbb17-1156-4615-83cb-bae935bda464.png)

### Why does the position matter?

**TL;DR:** The position of the Ulauncher matters because in my case the mouse gets moved to the currently focused window. Causing all programs to always be opened inside the left 'X-monitor' (even if the right 'X-Monitor' was focused).

**Long version:**
In this example, Ulauncher gets started while the right 'X-monitor' is focused. But Ulauncher opens at the center of the _hardware_ monitor, which by coincident is not inside the right 'X-monitor' but inside the left 'X-monitor'. Because my mouse pointer moves to the focused window on a focus change, it unfocuses the right 'X-monitor' and focuses the left 'X-monitor'. So if I now open an application, it starts inside the left 'X-monitor'. Even if I opened Ulauncher while the right 'X-monitor' was focused.

## Fix

I'm pretty sure that this is fixable without breaking the current behavior. (It works using Rofi).
I'll be using this issue to track the progress of the fix. 

It would be great to get some feedback from a maintainer if this fix would be welcome or not.

### Communication guidelines

- [X] I have read and agree with Ulauncher's communication guidelines and checked the [existing enhancement requests](https://github.com/Ulauncher/Ulauncher/labels/enhancement) before suggesting this.

**Comment by FinnHering** [comment]
Thank you for the directions! I'm going to attempt to fix. If it is an easy one, I might be able to pull it off this weekend. Will open a PR once I have a solution. :)

**Comment by FinnHering** [comment]
It seems like using GTK for this is impossible right not. See: https://gitlab.gnome.org/GNOME/gtk/-/issues/2013
This will make it hard to pull off without a library like python-xlib. 

But it seems like if I comment out position_window [here](https://github.com/Ulauncher/Ulauncher/blob/960607b3c8f9fbca70960c32e004b85e8e22cb09/ulauncher/ui/windows/UlauncherWindow.py#L245) it works like it should in all situations. Do you know what this call should do there?

**Comment by friday** [comment]
> @friday It seems like using GTK for this is impossible right not. See: https://gitlab.gnome.org/GNOME/gtk/-/issues/2013 This will make it hard to pull off without a library like python-xlib.

You could try poking around in Wnck, which is basically an X11 wrapper: https://lazka.github.io/pgi-docs/#Wnck-3.0/classes/Screen.html#Wnck.Screen

> But it seems like if I comment out position_window [here](https://github.com/Ulauncher/Ulauncher/blob/960607b3c8f9fbca70960c32e004b85e8e22cb09/ulauncher/ui/windows/UlauncherWindow.py#L245) it works like it should in all situations. Do you know what this call should do there?

This falls back on your Window manager default positioning. Either this will not center at all, or it centers both ways, so the window is constantly recentered when it changes height. We only want to to grow down.

> It's getting weirder

Would appreciate if you can keep trying to figure this out. I don't even know why things were done in a certain way/order to begin with, and I want to change a lot of this, but I generally don't mess with things unless it's broken or needs improvements.

**Comment by FinnHering** [comment]
> Would appreciate if you can keep trying to figure this out. I don't even know why things were done in a certain way/order to begin with, and I want to change a lot of this, but I generally don't mess with things unless it's broken or needs improvements.

I'm far from giving up! ;)

> You could try poking around in Wnck, which is basically an X11 wrapper: https://lazka.github.io/pgi-docs/#Wnck-3.0/classes/Screen.html#Wnck.Screen

That looks very interesting indeed! Thank you for the hint.

**Comment by friday** [comment]
Also, if we opened a new window each time we could get the "default" position, try to find out if this is centered and respect the x coordinate. But unfortunately the ulauncher window is being kept alive so it can be shown faster. And each time it is launched we have to assume the pointer is on a new location and re-run the positioning code.

One of my goals is to switch this around so that the window is opened fresh each time, because that's the only way to get it to center on the screen with the pointer if users are strictly on Wayland with no XWayland (and I also don't think the window needs to be kept in memory to launch fast, but this has yet to be proven).

**Comment by FinnHering** [comment]
That's true. Perhaps this is the easiest way. I've been looking at Wnck, but it doesn't provide any useful functionality to fix this issue. 
I'll try to cobble something together that allows the Ulauncher window to be removed from memory after it closes.

**Comment by friday** [comment]
> That's true. Perhaps this is the easiest way. I've been looking at Wnck, but it doesn't provide any useful functionality to fix this issue. I'll try to cobble something together that allows the Ulauncher window to be removed from memory after it closes.

That'd be a rather big change though. There's a lot of business logic there now that doesn't really belong there.

**Comment by FinnHering** [comment]
I agree. But I don’t really see another sane way. I could use libX11 like in the xinit.py but that is really hacky. Or I’d need to add some dependency that allows me to talk to X11. Another solution would be to downgrade to GTK2 and that comes with its own problems.
